### PR TITLE
Setup docker-compose healthcheck for MinIO

### PR DESCRIPTION
**Pull request by 0lmi**
Changelog: None
Ticket: QA-421
Signed-off-by: Alex Miliukov <alex@northern.tech>

As discussed recently, now, when we have more stable pipelines, it makes sense to get this improvement merged.

Related to the following tickets:
- https://tracker.mender.io/browse/QA-434
- https://tracker.mender.io/browse/QA-421

### CODE REVIEW

File: docker-compose.storage.minio.yml [hunk 1]
```diff
@@ -16,12 +16,18 @@ services:
         environment:
             MINIO_HTTP_TRACE: /dev/stdout
         command: server /export
+        healthcheck:
+            test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
+            interval: 1s
```
[the remark this answers is not in the document]
Comment (0lmi) [reply docker-compose.storage.minio.yml:21]: I actually have played with that a little bit (see pushes in https://github.com/mendersoftware/integration/pull/2053 for details) and this combination of interval, timeout and retries looked stable. But if to be honest it's hard for me to understand where 30s comes from...

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline please 🙏

File: docker-compose.storage.minio.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (0lmi) [reply docker-compose.storage.minio.yml:21]: I tested it locally and the best result is with `interval: 1s` and `start_period: 1s`. In this case locally it starts in the rage between 1.5s and 4s. How it works is it waits for `start_period` and then executes the check and repeats it with the `interval`. With `start_period: 5s` executed on not loaded server it will no difference between `interval: 10s` and `interval: 1s`, because very first check after `start_period` will pass, but on a loaded server with `interval: 10s` and failed 1st check it will not start faster than 15s. To me it looks like `start_period: 5s` + `interval: 1s` is good combination because locally it's only 1-2 secs slower that it can be in the best case, but in the same time in most cases very first check will pass and it won't generate any additional load (if executing HTTP GET by curl can be called so) even with `interval: 1s`. Not really sure what we win with 10s interval.

[the remark this answers is not in the document]
Comment (mzedel) [reply docker-compose.storage.minio.yml:21]: ah ok 👌 for me this was mainly about the recurring GET calls (where I think the recommendation was to keep the interval higher for less critical services to keep the load on the service down - but I guess that depends on the endpoint that's called to determine container health) - if that's not an issue at all 🚀 merge ahead!

**Comment by lluiscampos** [comment]
I have rebased it to get the latest stability improvements in tests. Let's see how it behaves now.

**Comment by lluiscampos** [comment]
@mender-test-bot start pipeline :ship:

**Comment by lluiscampos** [comment]
> I have rebased it to get the latest stability improvements in tests. Let's see how it behaves now.

It doesn't seem to work.

File: docker-compose.storage.minio.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (0lmi) [reply docker-compose.storage.minio.yml:21]: To make it in the best way I believe it would be nice to separate those settings for local runs/mender servers and our tests. For local runs it could be something like: `start_period: 10s` and `interval: 60s`, but for the tests the fastest feedback we get is better.

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

File: docker-compose.storage.minio.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ services:
     minio:
         # DO NOT upgrade this to any version "2021.05" or later.
         # We want to stay on Apache license for now.
-        image: minio/minio:RELEASE.2019-04-23T23-50-36Z
+        image: minio/minio:RELEASE.2021-04-22T15-44-28Z
```
Comment (lluiscampos) [inline docker-compose.storage.minio.yml:10]: This should be on a separate commit or mentioned in the commit body of the single commit.

File: docker-compose.storage.minio.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply docker-compose.storage.minio.yml:21]: I think we should not put this in `docker-compose.storage.minio.yml`, which may end up in production configuration for our users. Instead we should add it to `docker-compose.demo.yml` [here](https://github.com/mendersoftware/integration/blob/b3a0ad7e219bcf9d16896b9ceb5fa7baf4bf5553/docker-compose.demo.yml#L69), which will affect users only when running demo setup.

For testing, we use `docker-compose.demo.yml` with some [modifications](https://github.com/mendersoftware/integration/blob/b3a0ad7e219bcf9d16896b9ceb5fa7baf4bf5553/tests/run.sh#L89). Injecting there the `healthcheck` is very ugly.

**Review by lluiscampos — CHANGES_REQUESTED** [review]
Thank you very much for not giving up on this! :rocket:

**Review by lluiscampos — APPROVED** [review]

File: docker-compose.storage.minio.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (0lmi) [reply docker-compose.storage.minio.yml]: In general healthchecks are even more important for production systems. I don't mind to move it to `docker-compose.demo.yml` and leave `docker-compose.storage.minio.yml` without healthcheck to make some beauty (even considering that it's subjective judgment 🙃), considering that docker-compose based installation is unlikely can be called a real production one.

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

File: docker-compose.storage.minio.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lluiscampos) [reply docker-compose.storage.minio.yml]: > In general healthchecks are even more important for production systems. (...)

Good point. We could then have it in the base file with conservative intervals and override only the intervals to more aggressive ones for the `demo` setup.

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

File: tests/run.sh [hunk 3]
```diff
@@ -94,6 +94,8 @@ function modify_services_for_testing() {
     sed -e 's/DOWNLOAD_SPEED/#DOWNLOAD_SPEED/' -i ../docker-compose.testing.yml
     # whitelist *all* IPs/DNS names in the gateway (will be accessed via dynamically assigned IP in tests)
     sed -e 's/ALLOWED_HOSTS: .*/ALLOWED_HOSTS: ~./' -i ../docker-compose.testing.yml
+    # make sure that MinIO healthcheck from docker-compose.demo.yml is used independently from compose files sequence
+    sed -e '/^[[:space:]]*healthcheck\|test\|interval\|timeout\|retries\|start_period/d' -i ../docker-compose.storage.minio.yml
```
Comment (lluiscampos) [inline tests/run.sh:98]: Although I see your reasoning (good commit message, by the way!) it is very inconvenient for local development to edit a file that is included in git. After running a test this modification will be on the way for my local git status.

As you (kind of) said in the commit message, both in `demo` script and the test code we are placing the demo (or testing) yml file after `docker-compose.storage.minio.yml`, and these should be the only external entry points for users, tools, ci, etc.

`docker-compose.demo.yml` overrides other keys for other services in `docker-compose.yml` so it is already expected to be the last file added into the composition.

So please drop this commit :)

I personally don't think it is necessary, but one thing we could do for future reference is to add a comment in [docker_compose_manager.py](https://github.com/mendersoftware/integration/blob/b3a0ad7e219bcf9d16896b9ceb5fa7baf4bf5553/testutils/infra/container_manager/docker_compose_manager.py#L33-L34) to explicitly say that the order there is important.

**Review by lluiscampos — APPROVED** [review]
Thank you for splitting the style change into a separate commit :+1: 

Please drop the commit editing `docker-compose.storage.minio.yml` from `run.sh`, see reasoning in the comment below.

Otherwise good to go :guitar:

**Comment by 0lmi** [comment]
@mender-test-bot start pipeline

**Review by lluiscampos — APPROVED** [review]
